### ISSUE TITLE
command/shell modules incorrectly parse arguments when a k=v is present

### ISSUE THREAD

**Issue by timofei-durakov** [ansible/ansible#13862]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 2.0.0.1
##### Ansible Configuration:

no external changes
##### Environment:

Ubuntu 14.04
##### Summary:

Ansible incorrectly parses arumments in ad-hoc mode
##### Steps To Reproduce:

example: 

```
ansible all --sudo -f 5 -i inventory -m shell -a "mount -t nfs4 -o proto=tcp,port=2049 192.168.102.11:/ /opt/stack/data/nova/instances/"
```
##### Expected Results:

command succeed
##### Actual Results:

```
192.168.102.12 | FAILED | rc=0 >>
unsupported parameter for module: proto
```

there is a workaround by escaping '=' with '\\'

**Comment by timofei-durakov** [comment]
another case for this issue is base64 string containing '=' symbol

**Comment by laupow** [comment]
I can't get the `--become-user` option to work with ad hoc commands. I believe this may have also been an issue before Ansible 2, but it is still an issue in 2.0.0.2.

```
$ ansible all -i hosts-development -a "whoami" --become-user=root --ask-become-pass
SUDO password: 
myhost.com | SUCCESS | rc=0 >>
laupow
```

**Comment by bcoca** [comment]
@laupow you are missing a `--become`

**Comment by laupow** [comment]
Thanks @bcoca. 

I couldn't find `--become` mentioned in `man ansible` or on https://docs.ansible.com/ansible/intro_adhoc.html. I figured the other args implied that functionality when provided. 

Maybe the docs need an update? However it appears I may not be best suited to make changes.

**Comment by bcoca** [comment]
ansible --help
man ansible
http://docs.ansible.com/ansible/become.html

and in other places, I don't think everything needs to be in intro, though i might change the sudo examples there to become examples. (--become is equivalent to --sudo).

**Comment by gaieges** [comment]
Looks like this bug (and workaround) exists in 2.0.0.2 as well:

``` Shell
$ ansible --version
ansible 2.0.0.2
  config file = ~/.ansible.cfg
  configured module search path = Default w/o overrides
$ ansible -i inventory all --become -m shell -a 'docker ps -f name=consul'
HOST1 | FAILED | rc=0 >>
unsupported parameter for module: name
$ ansible -i inventory all --become -m shell -a 'docker ps -f name\=consul'
HOST1 | SUCCESS | rc=0 >>
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

HOST2 | SUCCESS | rc=0 >>
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS                                                                                                        NAMES
75f5ad637eb0        progrium/consul:latest   "/bin/start -server -"   23 minutes ago      Up 23 minutes       53/tcp, 0.0.0.0:8400->8400/tcp, 8300-8302/tcp, 8301-8302/udp, 0.0.0.0:8500->8500/tcp, 0.0.0.0:8600->53/udp   consul

HOST3 | SUCCESS | rc=0 >>
CONTAINER ID        IMAGE                    COMMAND                  CREATED             STATUS              PORTS                                                                                                        NAMES
536069d95441        progrium/consul:latest   "/bin/start -server -"   31 minutes ago      Up 31 minutes       53/tcp, 0.0.0.0:8400->8400/tcp, 8300-8302/tcp, 8301-8302/udp, 0.0.0.0:8500->8500/tcp, 0.0.0.0:8600->53/udp   consul


```

**Comment by pgporada** [comment]
I'm affected by this on Ansible 2.0.0.2

```
ansible vagrants -b -m shell -a "package-cleanup --oldkernels --count=2"
```

```
swarm-node02 | FAILED | rc=0 >>
unsupported parameter for module: --count
```

The workaround (escaping the =) does make it all work though.

```
ansible vagrants -b -m shell -a "package-cleanup --oldkernels --count\=2"
```

```
swarm-node02 | SUCCESS | rc=0 >>
Loaded plugins: fastestmirror, versionlock
No old kernels to remove
```

**Comment by jhutar** [comment]
Seen in ansible-2.0.1.0-2.el7.noarch as well. It is not present in ansible-1.9.4-1.el7.noarch.

Workaround (as mentioned by @pgporada) is not to use `--option=value`, but `--option value` (or escape equals sign).

**Comment by rkno82** [comment]
Is this issue, it's necessary to escape the equal sign, not fixed yet?

**Comment by alikins** [comment]
This seems to be fixed in current versions since devel commit:
```
commit ae21d98955f0d7367ea94610c14e66b78c7ad4ca
Author: James Cammarata <jimi@sngx.net>
Date:   Mon Mar 21 11:00:07 2016 -0400

    Properly use check_raw when using parse_kv in cli/ code
    
    Fixes ansible-modules-core#3070
```

The fix first appears in v2.0.2.0-1:

```
[fedora-25:ansible (v2.0.1.0-1-branch %)]$ ansible --version
ansible 2.0.1.0 (detached HEAD v2.0.1.0-1) last updated 2017/02/13 11:00:12 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 01f9b7262d) last updated 2017/02/13 10:49:32 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD e681ef8de3) last updated 2017/02/13 10:49:29 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-25:ansible (v2.0.1.0-1-branch %)]$ ansible localhost -m shell -a 'echo_argv ps -f name=consul -o dest=sdfsdf,src=sdfsd'
localhost | FAILED | rc=0 >>
unsupported parameter for module: dest

[fedora-25:ansible (v2.0.1.0-1-branch %)]$ git checkout v2.0.2.0-1-branch 
M       lib/ansible/modules/core
M       lib/ansible/modules/extras
Switched to branch 'v2.0.2.0-1-branch'
[fedora-25:ansible (v2.0.2.0-1-branch %)]$ ansible --version
ansible 2.0.2.0 (detached HEAD v2.0.2.0-1) last updated 2017/02/13 11:00:33 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 01f9b7262d) last updated 2017/02/13 10:49:32 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD e681ef8de3) last updated 2017/02/13 10:49:29 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-25:ansible (v2.0.2.0-1-branch %)]$ ansible localhost -m shell -a 'echo_argv ps -f name=consul -o dest=sdfsdf,src=sdfsd'
localhost | SUCCESS | rc=0 >>
ps
-f
name=consul
-o
dest=sdfsdf,src=sdfsd
```

The other stable branches have the fix as well (stable-2.0, stable-2.1, stable-2.2) and devel.

```
[fedora-25:ansible (devel % u=)]$ ansible --version              
ansible 2.3.0 (devel 8adb108aa9) last updated 2017/02/13 10:50:31 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-25:ansible (devel % u=)]$ ansible localhost -m shell -a 'echo_argv ps -f name=consul -o dest=sdfsdf,src=sdfsd'
 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS | rc=0 >>
ps
-f
name=consul
-o
dest=sdfsdf,src=sdfsd


[fedora-25:ansible (stable-2.2 % u=)]$ ansible --version
ansible 2.2.1.0 (stable-2.2 6176c95838) last updated 2017/02/13 10:56:50 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 01f9b7262d) last updated 2017/02/13 10:49:32 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD e681ef8de3) last updated 2017/02/13 10:49:29 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-25:ansible (stable-2.2 % u=)]$ ansible localhost -m shell -a 'echo_argv ps -f name=consul -o dest=sdfsdf,src=sdfsd'
 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS | rc=0 >>
ps
-f
name=consul
-o
dest=sdfsdf,src=sdfsd


[fedora-25:ansible (stable-2.1 % u=)]$ ansible --version
ansible 2.1.4.0 (stable-2.1 ad7c55dd9d) last updated 2017/02/13 11:01:57 (GMT -400)
  lib/ansible/modules/core: (detached HEAD 01f9b7262d) last updated 2017/02/13 10:49:32 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD e681ef8de3) last updated 2017/02/13 10:49:29 (GMT -400)
  config file = /home/adrian/.ansible.cfg
  configured module search path = Default w/o overrides
[fedora-25:ansible (stable-2.1 % u=)]$ ansible localhost -m shell -a 'echo_argv ps -f name=consul -o dest=sdfsdf,src=sdfsd'
 [WARNING]: provided hosts list is empty, only localhost is available

localhost | SUCCESS | rc=0 >>
ps
-f
name=consul
-o
dest=sdfsdf,src=sdfsd
```
